### PR TITLE
refactor(search): change git_url to source_id

### PR DESCRIPTION
**Pull request by xxs-wallace**
one of https://github.com/TabbyML/tabby/issues/2783

this PR is try to refactor code query to use source id instead of git url.

### CODE REVIEW

File: crates/tabby-common/src/index/code/mod.rs [hunk 1]
```diff
@@ -70,7 +72,7 @@ pub fn code_search_query(
 ) -> BooleanQuery {
     let schema = IndexSchema::instance();
     let corpus_query = schema.corpus_query(corpus::CODE);
-    let git_url_query = git_url_query(&query.git_url);
+    let git_url_query = git_url_query(&query.source_id);
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index/code/mod.rs:75]: Need to replace `git_url_query` to `source_query` to match against the `source_id` field in index:

https://demo.tabbyml.com/files/github/TabbyML/tabby/-/blob/48d9c084cc80d20bd439102cc0768920f6680e91/crates/tabby-common/src/index/mod.rs?plain=1#L102

File: crates/tabby-index/src/code/mod.rs [hunk 2]
```diff
@@ -85,7 +85,7 @@ impl IndexAttributeBuilder<SourceCode> for CodeBuilder {
             for await (start_line, body) in CodeIntelligence::chunks(&text, &source_code.language) {
                 let attributes = json!({
                     code::fields::CHUNK_FILEPATH: source_code.filepath,
-                    code::fields::CHUNK_GIT_URL: source_code.git_url,
+                    code::fields::CHUNK_SOURCE_ID: source_code.source_id,
```
Comment (wsxiaoys) [inline crates/tabby-index/src/code/mod.rs:88]: no need to add this, source_id is already indexed:

https://demo.tabbyml.com/files/github/TabbyML/tabby/-/blob/48d9c084cc80d20bd439102cc0768920f6680e91/crates/tabby-common/src/index/mod.rs?plain=1#L25

File: crates/tabby-index/src/code/types.rs [hunk 3]
```diff
@@ -10,7 +10,7 @@ use crate::indexer::{IndexId, ToIndexId};
 #[derive(Serialize, Deserialize, Clone)]
 pub struct SourceCode {
     pub id: String,
-    pub git_url: String,
+    pub source_id: String,
```
Comment (wsxiaoys) [inline crates/tabby-index/src/code/types.rs:13]: The refactor is to query against the index field `source_id`, there's no need to change field name to `source_id` every where.

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: crates/tabby-common/src/api/code.rs [hunk 4]
```diff
@@ -58,7 +58,7 @@ pub enum CodeSearchError {
 
 #[derive(Deserialize, ToSchema)]
 pub struct CodeSearchQuery {
-    pub git_url: String,
+    pub source_id: String,
```
Comment (wsxiaoys) [inline crates/tabby-common/src/api/code.rs:61]:
```suggestion
    pub git_url: String,
```

Given that we're also passing `git_url` from frontend (in `answer.rs`), we can keep the field name for simplicity.

File: crates/tabby-index/src/code/mod.rs (hunk 2 above)
Comment (xxs-wallace) [reply crates/tabby-index/src/code/mod.rs:88]: But it is attribute, not index.
Do you mean that we can just remote this field `CHUNK_GIT_URL`?

File: crates/tabby-common/src/index/code/mod.rs [hunk 5]
```diff
@@ -69,21 +61,10 @@ pub fn code_search_query(
     chunk_tokens_query: Box<dyn Query>,
 ) -> BooleanQuery {
     let schema = IndexSchema::instance();
-    let corpus_query = schema.corpus_query(corpus::CODE);
-    let git_url_query = git_url_query(&query.git_url);
+    let mut subqueries = schema.source_sub_queries(corpus::CODE, &query.git_url);
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index/code/mod.rs:64]:
```suggestion
    let mut subqueries = vec![
      (Occur::Must, Box::new(ConstScoreQuery::new(schema.source_query())))
    ]
```

No need to extract `source_sub_queries`? it probably leaks the implementation details.

File: crates/tabby/src/services/completion/completion_prompt.rs [hunk 6]
```diff
@@ -182,7 +182,7 @@ async fn collect_snippets(
     content: &str,
 ) -> Vec<Snippet> {
     let query = CodeSearchQuery {
-        git_url: git_url.to_owned(),
+        git_url: RepositoryConfig::canonicalize_url(git_url),
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion/completion_prompt.rs:185]: CodeSearch already handles canonicalization / closest_match internally, no need to make this change.

https://demo.tabbyml.com/files/github/TabbyML/tabby/-/blob/48d9c084cc80d20bd439102cc0768920f6680e91/crates/tabby/src/services/code.rs?plain=1#L78

File: ee/tabby-webserver/src/service/answer.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/answer.rs:1]: revert changes?

**Comment by wsxiaoys** [comment]
Please also fix all CI checks

File: crates/tabby-common/src/api/code.rs (hunk 4 above)
Comment (xxs-wallace) [reply crates/tabby-common/src/api/code.rs]: ok

File: crates/tabby-common/src/index/code/mod.rs (hunk 5 above)
Comment (xxs-wallace) [reply crates/tabby-common/src/index/code/mod.rs:64]: Yes. But the origin `source_query` has add `corpus_query` beside it. I just keep the logic like creating `source_query`

File: ee/tabby-webserver/src/service/answer.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/answer.rs:1]: revert changes?

File: crates/tabby-common/src/index/code/mod.rs (hunk 5 above)
Comment (xxs-wallace) [reply crates/tabby-common/src/index/code/mod.rs:64]: Although all query are composed with `Must`, I just want to keep the same logic with previous code.

Comment (wsxiaoys) [reply crates/tabby-common/src/index/code/mod.rs:64]: I don't get it - nested must is still must? it doesn't change the logic if with my suggested change.

**Review by wsxiaoys — APPROVED** [review]

**Pull request merged**